// @node-red/editor-api/lib/admin/semver.js
const apiUtil = require("../util");

let runtimeAPI;

module.exports = {
    init: function(_runtimeAPI) {
        runtimeAPI = _runtimeAPI;
    },
    checkEngines: function(req, res) {
        const opts = {
            user: req.user,
            engines: req.body,
        };

        runtimeAPI.semver.checkEngines(opts)
            .then((result) => res.json(result))
            .catch((error) => apiUtil.rejectHandler(req, res, error));
    },
    compare: function(req, res) {
        const opts = {
            user: req.user,
            v1: req.body.v1,
            v2: req.body.v2,
        };

        runtimeAPI.semver.compare(opts)
            .then((result) => res.json(result))
            .catch((error) => apiUtil.rejectHandler(req, res, error));
    },
    satisfies: function(req, res) {
        const opts = {
            user: req.user,
            version: req.body.version,
            range: req.body.range,
        };

        runtimeAPI.semver.satisfies(opts)
            .then((result) => res.json(result))
            .catch((error) => apiUtil.rejectHandler(req, res, error));
    },
};
